File: aether/backend/app/tools/audit_engine.py
import asyncio
import hashlib
import json
import re
from typing import Any, Dict, List
from urllib.parse import parse_qsl, urlencode, urlparse, urlunparse

import requests

SQLI_PAYLOADS = ["'", '"', "OR 1=1"]
SQL_ERROR_PATTERNS = [
    re.compile(pattern, re.IGNORECASE)
    for pattern in (
        r"sql syntax",
        r"mysql",
        r"postgresql",
        r"sqlite",
        r"odbc",
        r"sqlstate",
        r"unterminated quoted string",
        r"syntax error at or near",
    )
]
RATE_LIMIT_HEADERS = {
    "x-ratelimit-limit",
    "x-ratelimit-remaining",
    "x-ratelimit-reset",
    "ratelimit-limit",
    "ratelimit-remaining",
    "ratelimit-reset",
}
KNOWN_WAF_SIGNATURES = {
    "cloudflare": "Cloudflare",
    "akamai": "Akamai",
    "sucuri": "Sucuri",
    "imperva": "Imperva",
    "fastly": "Fastly",
    "aws": "AWS",
}
EXPOSED_SERVER_PATTERNS = [
    re.compile(pattern, re.IGNORECASE)
    for pattern in (
        r"apache/\d",
        r"nginx/\d",
        r"iis/\d",
        r"php/\d",
        r"express",
        r"gunicorn/\d",
        r"uvicorn",
    )
]


def _finding_id(prefix: str, *parts: str) -> str:
    digest = hashlib.sha1(":".join([prefix, *parts]).encode("utf-8")).hexdigest()[:12]
    return f"{prefix}:{digest}"


def _request_snapshot(target_url: str) -> Dict[str, Any]:
    response = requests.get(target_url, timeout=8, allow_redirects=True)
    return {
        "url": response.url,
        "status_code": response.status_code,
        "headers": {key.lower(): value for key, value in response.headers.items()},
        "text": response.text[:8000],
    }


def _profile_target(target_url: str, base_response: Dict[str, Any]) -> Dict[str, Any]:
    headers = base_response.get("headers", {})
    findings: List[Dict[str, Any]] = []
    profiles: List[Dict[str, Any]] = []

    rate_limit_present = any(header in headers for header in RATE_LIMIT_HEADERS)
    header_blob = json.dumps(headers).lower()
    waf_matches = [label for signature, label in KNOWN_WAF_SIGNATURES.items() if signature in header_blob]
    server_banner = headers.get("server", "")
    exposed_banner = bool(server_banner and any(pattern.search(server_banner) for pattern in EXPOSED_SERVER_PATTERNS))

    if not rate_limit_present:
        findings.append(
            {
                "id": _finding_id("vuln", "rate-limit", target_url),
                "category": "ddos_heuristic",
                "title": "Missing Rate-Limit Signaling",
                "severity": "medium",
                "detail": "No standard rate-limit headers were observed, which may indicate weak request-throttling controls.",
                "evidence": {"headers_seen": sorted(headers.keys())[:20]},
            }
        )

    if not waf_matches:
        findings.append(
            {
                "id": _finding_id("vuln", "waf", target_url),
                "category": "ddos_heuristic",
                "title": "No WAF Signature Detected",
                "severity": "low",
                "detail": "No common upstream WAF signature was detected in the response headers, reducing confidence in edge-layer request filtering.",
                "evidence": {"headers_seen": sorted(headers.keys())[:20]},
            }
        )

    if exposed_banner:
        findings.append(
            {
                "id": _finding_id("vuln", "server-banner", server_banner),
                "category": "ddos_heuristic",
                "title": "Verbose Server Banner",
                "severity": "medium",
                "detail": "The server banner exposes implementation details that can help attackers tune exploit and traffic-flood strategies.",
                "evidence": {"server": server_banner},
            }
        )

    profiles.append(
        {
            "profile_type": "vulnerability_profiler",
            "label": "Transport Resilience",
            "summary": "Profiles rate-limit signaling, edge filtering, and banner exposure for abuse resistance.",
            "details": {
                "rate_limit_present": rate_limit_present,
                "waf_signatures": waf_matches,
                "server_banner": server_banner or None,
                "exposed_banner": exposed_banner,
            },
        }
    )

    return {"findings": findings, "profiles": profiles}


def _test_sqli_reflections(target_url: str) -> Dict[str, Any]:
    parsed = urlparse(target_url)
    params = parse_qsl(parsed.query, keep_blank_values=True)
    findings: List[Dict[str, Any]] = []

    if not params:
        return {"tested_params": [], "findings": []}

    tested_params: List[str] = []

    for index, (key, value) in enumerate(params):
        for payload in SQLI_PAYLOADS:
            mutated_params = list(params)
            mutated_params[index] = (key, f"{value}{payload}")
            mutated_url = urlunparse(parsed._replace(query=urlencode(mutated_params, doseq=True)))
            tested_params.append(f"{key}:{payload}")

            try:
                response = _request_snapshot(mutated_url)
            except Exception:
                continue

            text = response.get("text", "")
            has_sql_error = any(pattern.search(text) for pattern in SQL_ERROR_PATTERNS)
            payload_reflected = payload in text

            if has_sql_error:
                findings.append(
                    {
                        "id": _finding_id("vuln", "sqli-error", key, payload),
                        "category": "sqli_heuristic",
                        "title": f"SQL Error Reflection via {key}",
                        "severity": "high",
                        "detail": f"The parameter '{key}' triggered a database-flavored error response when probed with a minimal SQLi heuristic payload.",
                        "evidence": {
                            "parameter": key,
                            "payload": payload,
                            "status_code": response.get("status_code"),
                            "url": response.get("url"),
                        },
                    }
                )
            elif payload_reflected:
                findings.append(
                    {
                        "id": _finding_id("vuln", "reflection", key, payload),
                        "category": "sqli_heuristic",
                        "title": f"Reflected Input Surface via {key}",
                        "severity": "medium",
                        "detail": f"The parameter '{key}' reflected a probe payload in the response body, which warrants deeper input-handling review.",
                        "evidence": {
                            "parameter": key,
                            "payload": payload,
                            "status_code": response.get("status_code"),
                            "url": response.get("url"),
                        },
                    }
                )

    return {"tested_params": tested_params, "findings": findings}


def _run_audit(target_url: str) -> Dict[str, Any]:
    base_response = _request_snapshot(target_url)
    sqli_result = _test_sqli_reflections(target_url)
    profile_result = _profile_target(target_url, base_response)

    return {
        "target_url": target_url,
        "tested_params": sqli_result["tested_params"],
        "base_response": {
            "url": base_response.get("url"),
            "status_code": base_response.get("status_code"),
            "headers": base_response.get("headers"),
        },
        "findings": [*sqli_result["findings"], *profile_result["findings"]],
        "profiles": profile_result["profiles"],
    }


async def audit_engine(target_url: str) -> Dict[str, Any]:
    try:
        return await asyncio.to_thread(_run_audit, target_url)
    except Exception as error:
        return {
            "target_url": target_url,
            "tested_params": [],
            "base_response": {},
            "findings": [],
            "profiles": [],
            "error": f"Audit engine failed: {error}",
        }


def format_audit_logs(audit_result: Dict[str, Any]) -> List[str]:
    if audit_result.get("error"):
        return [f"[EXECUTE] AUDIT_ENGINE: {audit_result['error'].upper()}."]

    findings = audit_result.get("findings", [])
    tested_params = audit_result.get("tested_params", [])
    profiles = audit_result.get("profiles", [])
    logs = [
        f"[EXECUTE] AUDIT_ENGINE: PROFILED {len(tested_params)} PARAMETER PROBES AND {len(profiles)} RESILIENCE PROFILES."
    ]

    if findings:
        summary = "; ".join(
            f"{finding.get('category', 'finding').upper()}::{finding.get('title', 'UNKNOWN')[:42].upper()}"
            for finding in findings[:3]
        )
        logs.append(f"[EXECUTE] AUDIT_ENGINE: VULNERABILITY SIGNALS DETECTED - {summary}.")
    else:
        logs.append("[EXECUTE] AUDIT_ENGINE: NO SQLI OR ABUSE-RESILIENCE HEURISTICS TRIPPED DURING THIS HUNT.")

    return logs

File: aether/src/lib/scanner/tools/audit-engine.ts
import { createHash } from "crypto";
import type { AuditResult, Finding } from "../types";

const SQLI_PAYLOADS = ["'", '"', "OR 1=1"];
const SQL_ERROR_PATTERNS = [
  /sql syntax/gi, /mysql/gi, /postgresql/gi, /sqlite/gi,
  /odbc/gi, /sqlstate/gi, /unterminated quoted string/gi,
  /syntax error at or near/gi,
];
const RATE_LIMIT_HEADERS = new Set([
  "x-ratelimit-limit", "x-ratelimit-remaining", "x-ratelimit-reset",
  "ratelimit-limit", "ratelimit-remaining", "ratelimit-reset",
]);
const KNOWN_WAF_SIGNATURES: Record<string, string> = {
  cloudflare: "Cloudflare", akamai: "Akamai", sucuri: "Sucuri",
  imperva: "Imperva", fastly: "Fastly", aws: "AWS",
};
const EXPOSED_SERVER_PATTERNS = [
  /apache\/\d/i, /nginx\/\d/i, /iis\/\d/i, /php\/\d/i,
  /express/i, /gunicorn\/\d/i, /uvicorn/i,
];

function findingId(prefix: string, ...parts: string[]): string {
  const digest = createHash("sha1").update(parts.join(":")).digest("hex").slice(0, 12);
  return `${prefix}:${digest}`;
}

async function requestSnapshot(targetUrl: string) {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 8000);
  try {
    const response = await fetch(targetUrl, {
      signal: controller.signal,
      redirect: "follow",
      headers: { "User-Agent": "AETHER-Security-Auditor/1.0" },
    });
    clearTimeout(timeout);
    const headers: Record<string, string> = {};
    response.headers.forEach((v, k) => { headers[k.toLowerCase()] = v; });
    const text = (await response.text()).slice(0, 8000);
    return { url: response.url, status_code: response.status, headers, text };
  } catch (error) {
    clearTimeout(timeout);
    throw error;
  }
}

function profileTarget(targetUrl: string, baseResponse: { headers: Record<string, string>; status_code: number; url: string }) {
  const headers = baseResponse.headers;
  const findings: Finding[] = [];
  const profiles: AuditResult["profiles"] = [];

  const rateLimitPresent = Object.keys(headers).some((h) => RATE_LIMIT_HEADERS.has(h));
  const headerBlob = JSON.stringify(headers).toLowerCase();
  const wafMatches = Object.entries(KNOWN_WAF_SIGNATURES)
    .filter(([sig]) => headerBlob.includes(sig))
    .map(([, label]) => label);
  const serverBanner = headers["server"] || "";
  const exposedBanner = serverBanner && EXPOSED_SERVER_PATTERNS.some((p) => p.test(serverBanner));

  if (!rateLimitPresent) {
    findings.push({
      id: findingId("vuln", "rate-limit", targetUrl),
      category: "ddos_heuristic",
      title: "Missing Rate-Limit Signaling",
      severity: "Medium",
      detail: "No standard rate-limit headers were observed, which may indicate weak request-throttling controls.",
      attack_vector: "Response header inspection",
      detected_threat: "Missing rate-limit headers",
      evidence_snippet: JSON.stringify(Object.keys(headers).slice(0, 20)),
      provided_solution: "Add rate-limiting headers to API responses.",
      evidence: { headers_seen: Object.keys(headers).slice(0, 20) },
    });
  }
  if (wafMatches.length === 0) {
    findings.push({
      id: findingId("vuln", "waf", targetUrl),
      category: "ddos_heuristic",
      title: "No WAF Signature Detected",
      severity: "Low",
      detail: "No common upstream WAF signature was detected in the response headers.",
      attack_vector: "Response header inspection",
      detected_threat: "No WAF signature",
      evidence_snippet: JSON.stringify(Object.keys(headers).slice(0, 20)),
      provided_solution: "Consider deploying a WAF in front of the application.",
      evidence: { headers_seen: Object.keys(headers).slice(0, 20) },
    });
  }
  if (exposedBanner) {
    findings.push({
      id: findingId("vuln", "server-banner", serverBanner),
      category: "ddos_heuristic",
      title: "Verbose Server Banner",
      severity: "Medium",
      detail: "The server banner exposes implementation details that can help attackers tune exploit strategies.",
      attack_vector: "Response header inspection",
      detected_threat: `Server banner: ${serverBanner}`,
      evidence_snippet: serverBanner,
      provided_solution: "Remove or obfuscate the Server header.",
      evidence: { server: serverBanner },
    });
  }

  profiles.push({
    profile_type: "vulnerability_profiler",
    label: "Transport Resilience",
    summary: "Profiles rate-limit signaling, edge filtering, and banner exposure for abuse resistance.",
    details: {
      rate_limit_present: rateLimitPresent,
      waf_signatures: wafMatches,
      server_banner: serverBanner || null,
      exposed_banner: !!exposedBanner,
    },
  });

  return { findings, profiles };
}

async function testSqlReflections(targetUrl: string): Promise<{ tested_params: string[]; findings: Finding[] }> {
  let parsed: URL;
  try {
    parsed = new URL(targetUrl);
  } catch {
    return { tested_params: [], findings: [] };
  }

  const params = Array.from(parsed.searchParams.entries());
  if (params.length === 0) return { tested_params: [], findings: [] };

  const findings: Finding[] = [];
  const testedParams: string[] = [];

  for (const [key, value] of params) {
    for (const payload of SQLI_PAYLOADS) {
      testedParams.push(`${key}:${payload}`);
      const mutatedUrl = new URL(parsed.toString());
      mutatedUrl.searchParams.set(key, `${value}${payload}`);

      try {
        const result = await Promise.race([
          requestSnapshot(mutatedUrl.toString()),
          new Promise<never>((_, reject) => setTimeout(() => reject(new Error("timeout")), 5000)),
        ]);
        const text = result.text || "";
        const hasSqlError = SQL_ERROR_PATTERNS.some((p) => p.test(text));
        const payloadReflected = text.includes(payload);

        if (hasSqlError) {
          findings.push({
            id: findingId("vuln", "sqli-error", key, payload),
            category: "sqli_heuristic",
            title: `SQL Error Reflection via ${key}`,
            severity: "High",
            detail: `The parameter '${key}' triggered a database-flavored error response.`,
            attack_vector: "SQL injection payload injection",
            detected_threat: `SQL error from parameter: ${key}`,
            evidence_snippet: `Payload: ${payload}, Status: ${result.status_code}`,
            provided_solution: "Use parameterized queries for all database interactions.",
            evidence: { parameter: key, payload, status_code: result.status_code, url: result.url },
          });
        } else if (payloadReflected) {
          findings.push({
            id: findingId("vuln", "reflection", key, payload),
            category: "sqli_heuristic",
            title: `Reflected Input Surface via ${key}`,
            severity: "Medium",
            detail: `The parameter '${key}' reflected a probe payload in the response body.`,
            attack_vector: "Input reflection testing",
            detected_threat: `Reflected payload in parameter: ${key}`,
            evidence_snippet: `Payload: ${payload}, Status: ${result.status_code}`,
            provided_solution: "Review input handling and apply contextual output encoding.",
            evidence: { parameter: key, payload, status_code: result.status_code, url: result.url },
          });
        }
      } catch {
        continue;
      }
    }
  }

  return { tested_params: testedParams, findings };
}

export async function auditEngine(targetUrl: string): Promise<AuditResult> {
  try {
    const baseResponse = await requestSnapshot(targetUrl);
    const sqliResult = await testSqlReflections(targetUrl);
    const profileResult = profileTarget(targetUrl, baseResponse);

    return {
      target_url: targetUrl,
      tested_params: sqliResult.tested_params,
      base_response: {
        url: baseResponse.url,
        status_code: baseResponse.status_code,
        headers: baseResponse.headers,
      },
      findings: [...sqliResult.findings, ...profileResult.findings],
      profiles: profileResult.profiles,
    };
  } catch (error) {
    return {
      target_url: targetUrl,
      tested_params: [],
      base_response: {},
      findings: [],
      profiles: [],
      error: `Audit engine failed: ${error}`,
    };
  }
}

export function formatAuditLogs(result: AuditResult): string[] {
  if (result.error) return [`[EXECUTE] AUDIT_ENGINE: ${result.error.toUpperCase()}.`];
  const logs = [
    `[EXECUTE] AUDIT_ENGINE: PROFILED ${result.tested_params.length} PARAMETER PROBES AND ${result.profiles.length} RESILIENCE PROFILES.`,
  ];
  if (result.findings.length > 0) {
    const summary = result.findings.slice(0, 3)
      .map((f) => `${f.category.toUpperCase()}::${f.title.toUpperCase().slice(0, 42)}`)
      .join("; ");
    logs.push(`[EXECUTE] AUDIT_ENGINE: VULNERABILITY SIGNALS DETECTED - ${summary}.`);
  } else {
    logs.push("[EXECUTE] AUDIT_ENGINE: NO SQLI OR ABUSE-RESILIENCE HEURISTICS TRIPPED DURING THIS HUNT.");
  }
  return logs;
}
